gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/P.java

```diff
@@ -74,7 +77,16 @@ public void setValue(final V value) {
 
     @Override
     public boolean test(final V testValue) {
-        return this.biPredicate.test(testValue, this.value);
+        if (this.value instanceof GValue) {
```

Comment:
I think this piece of code will be repeated many times, it is better to make a static method somewhere

Reply:
I have updated this to use `GValue.valueOf()` and have made similar changes where applicable throughout the branch.